"Marginal Markets" addition to price node

rodvar opened the pull request:
## Purpose

 - Integrate "marginal markets" into Bisq 1&2
 - Implementation for:

1. https://github.com/bisq-network/bisq2/issues/2645
2. https://github.com/bisq-network/bisq2/issues/2230
3. https://github.com/bisq-network/bisq2/issues/2148
4. https://github.com/bisq-network/bisq2/issues/1952

## Approach

 - Yadio free public API was proposed by a Bisq user. It gives the BTC price in USD and all the USD prices in different currencies, including the ones we are interested in (see issues listed above) and more.
  - It also provide LBP (Lebanon Pound) price that seems accurate (blue rate) as well as another pricepoint for the current ARS support, which is included in this PR (note the LBP will be removed from the banned currencies list after this gets merged)
  - A whitelist is provided to control which yadio currencies we want to accept. Initial proposal is the 4 currencies requested above + ARS + LBP (blue market prices)
 
 ## How to test this PR
 
  - An integration test YadioTest is provided
  - Also run the pricenode and use the local APIs as explained in this project README.md

##  Dev logs

 - Implementation of Yadio provider for marginal currencies 
- implemented whitelist for more control over added currencies
 - whitelisted current bisq issues requested currencies + blue market currencies (redundancy for ARS and FIRST PRICE for LBP)
 - added docs to the main class
 - test provided
 - gitignore bash scripts

rodvar committed:
- Implementation of Yadio provider for marginal currencies
 - implemented whitelist for more control over added currencies
 - whitelisted current bisq issues requested currencies + blue market
   currencies (redundancy for ARS and FIRST PRICE for LBP)
 - added docs to the main class
 - test provided
 - gitignore bash scripts

HenrikJannsen reviewed (approved):
utACK

devinbileck commented on src/main/java/bisq/price/spot/providers/Yadio.java:
```diff
@@ -0,0 +1,144 @@
+ *     <li>PYG (Paraguayan Guarani</li>
+ *     <li>DOP (Dominican Peso)</li>
+ *     <li>BOB (Bolivian Bolivariano</li>
+ *     <li>EGP (Egyptian Pound</li>
```
Nit, you are missing closing parenthesis on a few of these currency names.
```suggestion
 *     <li>EGP (Egyptian Pound)</li>
```

devinbileck commented on .gitignore:
```diff
@@ -4,4 +4,5 @@ build
 bisq-pricenode
 bisq-pricenode.bat
 lib
+*.sh
```
I am curious why you are ignoring all shell scripts?

rodvar replied:
I like to make my own shell scripts in the project root and since its not used in the project and its annoying to get git asking to commit them all the time

alvasw replied:
You can add this to your local `.gitignore` file (`$GIT_DIR/info/exclude`).
See: https://git-scm.com/docs/gitignore#_description

devinbileck commented on src/main/java/bisq/price/spot/providers/Yadio.java:
```diff
@@ -0,0 +1,144 @@
+ * Further analysis could be made to incorporate more currencies like VES (Venezuela).
+ */
+@Component
+public class Yadio extends ExchangeRateProvider implements BlueRateProvider {
```
Nit, can be made package private.

devinbileck commented on src/main/java/bisq/price/spot/providers/Yadio.java:
```diff
@@ -0,0 +1,144 @@
+@Component
+public class Yadio extends ExchangeRateProvider implements BlueRateProvider {
+
+    public static final String PROVIDER_NAME = "YADIO";
```
Nit, can be made private.

devinbileck commented on src/main/java/bisq/price/spot/providers/Yadio.java:
```diff
@@ -0,0 +1,144 @@
+        yadioTicker.setTimestamp((Long) yadioMarketData.get("timestamp"));
+        @SuppressWarnings("unchecked")
+        Map<String, Double> usdMap = (Map<String, Double>) yadioMarketData.getOrDefault("USD", Collections.emptyMap());
+        yadioTicker.setUSD(usdMap);
```
To me, the above attributes do not constitute a "ticker". Rather, it is a collection of exchange rates with the USD attribute containing the individual tickers for each exchange rate. Would be a good idea to update the naming accordingly.

Also, if you define an appropriate POJO perhaps it would be able to deserialize it without having to manually parse the data?

rodvar replied:
Tried to setup the POJO but couldn't get it to pickup the raw data + map of String,Double for the usd prices of different fiat.
If you wanna give it a go yourself you are most welcome to collaborate!

devinbileck commented on src/main/java/bisq/price/spot/providers/Yadio.java:
```diff
@@ -0,0 +1,144 @@
+                new ExchangeRate(
+                        currencySymbol,
+                        price,
+                        System.currentTimeMillis(),
```
Since the timestamp is included in the response, perhaps it would be more appropriate/accurate to use that value?

devinbileck reviewed (changes requested):
Aside from a few nits and suggestions, I am encountering a test failure in YadioTest when trying to build with `./gradlew clean build`:

```
org.opentest4j.AssertionFailedError: Retrieved exchange rates contain unsupported currencies: [LBP] ==> expected: <true> but was: <false>
```

Are you not encountering this failure?

rodvar commented:
> Aside from a few nits and suggestions, I am encountering a test failure in YadioTest when trying to build with `./gradlew clean build`:
> 
> ```
> org.opentest4j.AssertionFailedError: Retrieved exchange rates contain unsupported currencies: [LBP] ==> expected: <true> but was: <false>
> ```
> 
> Are you not encountering this failure?

Hey @devinbileck , thanks for taking the time to review this PR! I've fixed most of your suggestions/requests in the coming commit. 

Special thanks for this comment which I could no spot when doing gradlew test or running from the IDE - Thank you!!

I'll fix it as well as an old test issue that might occur or not (depending on where is the request coming from)

rodvar committed:
- fix crytoYa test failing depending on where your connection is
   comming from (looks like they restrict some countries IP addresses
now)
 - fix Excluded fiat ExchangeServiceTest setup affecting new yadio test
   when running from gradlew build
 - other minor suggestions from PR review

devinbileck reviewed (approved):
ACK
This is running on my price node devinpndvdwll4wiqcyq5e7itezmarg7rzicrvf6brzkwxdm374kmmyd.onion/getAllMarketPrices

alvasw reviewed (changes requested):
What is the reasoning behind adding `*.sh` to the `.gitignore` file?
The `docker/` and `scripts/` folders contain scripts (`.sh` files). This change makes it harder for others to edit and push there changes.

rodvar commented:
> What is the reasoning behind adding `*.sh` to the `.gitignore` file? The `docker/` and `scripts/` folders contain scripts (`.sh` files). This change makes it harder for others to edit and push there changes.

in that case I'll remove it from the PR

rodvar committed:
Update .gitignore

revert gitignore change

alvasw reviewed (approved):
utACK

Thank you very much!

rodvar commented:
@HenrikJannsen I think we are good to merge this one now :muscle:
